Refactor CI with separate deploy jobs

erikmd opened the pull request:
Dear @CohenCyril,
Following @ejgallego's suggestion I've carefully read the GitLab CI doc related to environments and in particular [this section](https://gitlab.com/help/ci/variables/README.md#limiting-environment-scopes-of-variables-premium).

Note: the docker push to Docker Hub will only work in `master`, in the `deployment` environment but I wasn't able to easily test this "environment" feature as it is limited to Premium accounts… (it doesn't work with *private* repos on gitlab.com) so I'm afraid we'll be able to really test whether the deploy still work only after the PR is merged in master…

I also added some `echo` commands and `git rev-parse --verify HEAD` to print more debugging info.

Summary of advantages/drawbacks:

* Pros:
  * as pointed out by Emilio, this new design is more safe as external code cloned by GItLab CI won't be able to read the environment variables specific to the `.docker-deploy` trusted jobs.
  * the semantics of the CI is slightly changed as `mathcomp-dev` for coq-(8.6, 8.7, 8.8, 8.9, dev) will be pushed to Docker Hub **only** if the build stage **and** the test stage were both successful. This is the breaking change of the PR, but I put it in the Pros list as it seems something desirable (?) 
* Cons:
  * the overall time of the CI may be a bit longer as there is an additional `docker pull "${IMAGE}"` command.

Note: I've just switched the **protected variables** related to Docker Hub to **scoped protected variables** using [this url](https://gitlab.com/math-comp/math-comp/settings/ci_cd) (scoped = specific to the `deployment` environment), so if another PR is merged in master before that PR, it won't be able to push to Docker Hub.

Feel free to comment if you have questions/remarks.

ejgallego commented:
I'd say it looks really good, congrats!

About testing, I'm ccing @vbgl and @SkySkimmer who are experts here and may know a bit better how they do test.

SkySkimmer commented:
>the semantics of the CI is slightly changed as mathcomp-dev for coq-(8.6, 8.7, 8.8, 8.9, dev) will be pushed to Docker Hub only if the build stage and the test stage were both successful. This is the breaking change of the PR, but I put it in the Pros list as it seems something desirable (?)

If you end up considering it a con you should be able to change it by using `when: always` https://docs.gitlab.com/ee/ci/yaml/#when

I have no expertise in deploy/environment.

erikmd commented:
@SkySkimmer 
> If you end up considering it a con you should be able to change it by using `when: always`

Thanks! this could be a useful work-around, but on the other hand, if a job from `build` with a given version of Coq fails:
* all other images would be pushed, which might be unwanted,
* and 2 jobs would be failing (e.g. coq-8.7 and mathcomp-dev:coq-8.7), which would clutter a bit coqbot's report.

So I end up implementing another solution (swapping the `deploy` and `test` stages) that should address these two drawbacks…

erikmd commented:
@CohenCyril I've noted that #307 now has a conflict so FYI I first rebased #306 on master and took the opportunity to improve details (removing unnecessary `=` in echo commands) and push a clone of the repo in a temporary public gitlab repo to test the new deployment procedure (temporarily to https://hub.docker.com/r/erikmd/math-comp-env-1). If this works I'll rebase #307 on #306 this evening and you'll be able to merge both PRs...

erikmd commented:
@CohenCyril it was a good idea to test the new deployment procedure, because it seems it didn't work out of the box :)
I identified two small issues (one regarding a spurious quote character, and one regarding the procedure itself, following @ejgallego's advice to rely on a deployment `environment`)
I fixed the first issue but I'll need more time to finish this on tomorrow; anyway this approach indeed seems promising.
See you, Erik

CohenCyril commented:
Thanks a lot @erikmd, waiting for your next input.

CohenCyril commented:
Hi! @erikmd any news?

erikmd commented:
Hi @CohenCyril ! sorry, I've not had enough time to debug/finalize the new CI deployment setup…
But this task is in my todo list for this week.

erikmd commented:
Hi @CohenCyril, I've finish upgrading the `coqorg/coq` images so they now rely on opam 2.0.4:
- https://hub.docker.com/r/coqorg/base
- https://hub.docker.com/r/coqorg/coq

and the `coqorg/coq:dev` images now contain some metadata to know more easily what is the underlying git commit hash of coq: https://microbadger.com/images/coqorg/coq:dev

Then I pursued my experiments on #306 and #307 and the deploy configuration seems very fine now…
i've rebased on `master` (to take into account the removal of `coq-8.6` jobs) and force-pushed so these two CI pipelines are now pending:
- https://gitlab.com/math-comp/math-comp/pipelines/57070048 (for #306)
- https://gitlab.com/math-comp/math-comp/pipelines/57070300 (for #307)

FYI I've switched again the protected variables related to Docker Hub to *scoped protected variable*s w.r.t. the deployment environment, which implies this PR should be merged in `master` **before** other PRs (so that upcoming merges can benefit from the deploy to <https://hub.docker.com/r/mathcomp/mathcomp-dev>).

I propose that you merge #306 then #307 when the corresponding pipelines will be green…

Kind regards, Erik

CohenCyril commented:
@erikmd CI status are not displayed in PRs anymore... (they did this morning AFAIK) would you know why?

erikmd commented:
@CohenCyril indeed, I do not why…
I checked the items from [this checklist](https://github.com/coq/bot#rely-on-coqbot-to-synchronize-your-github-prs-with-your-gitlab-project) (for the GitLab side) and saw nothing special…

@Zimmi48 do you have an idea of what could cause that lack of CI status reporting?

Zimmi48 commented:
Sorry I didn't see the issue. Can you link me to a PR where there is a problem?

CohenCyril commented:
> Sorry I didn't see the issue. Can you link me to a PR where there is a problem?

@Zimmi48 The problem lasted the whole afternoon, but now it seems solved: I cannot find a PR with this problem anymore...

Zimmi48 commented:
I had a look at the log. It looks like GitLab didn't send any webhooks during the afternoon, until 16:30 when a lot of delayed webhooks were emitted. Confirmed this is was a general GitLab problem: https://twitter.com/gitlabstatus/status/1118076245688737793

erikmd commented:
:+1: The CI for this first PR was green :)

erikmd commented:
And that of #307 was green as well; so I guess both PRs are ready.
Cc @CohenCyril

CohenCyril commented:
Thanks @erikmd !!
Do you know how to configure github to alert users that changes to this file might leak a secret variable and should hence be handled very carefully?

ejgallego commented:
@CohenCyril you can use a codeowners file so every commit touching gitlab-ci.yml requires a review; other than there is not further check.

Maybe a worthwhile feature request for gitlab would be to require a signed gitlab-ci.yml file

ejgallego commented:
Yes so indeed as @erikmd points out, the main purpose of the PR is that now the change leaking information would have to reach master, thus indeed committers to the main branch should be careful of course with changes to the CI setup.

CohenCyril commented:
@ejgallego 
> indeed committers to the main branch should be careful of course with changes to the CI setup.

this was my point

@erikmd 


> and finally I enabled a recently implemented protection in GitLab to mask the corresponding value if ever it were printed in the log:

wow ok, I did not know about this... very useful (but I guess it would only prevent mistakes, not deliberate hacking attempts, which could encore the token so that gitlab does not filter it out?)

Zimmi48 commented:
Cool new feature indeed! Unfortunately, we can't use it for our secret variables in the Coq project because they don't respect the acceptable format. So we'll have to continue relying on protected + environment restricted + being careful about not printing the environment in deployment jobs.

erikmd commented:
Just to complement @ejgallego's comment:
> the main purpose of the PR is that now the change leaking information would have to reach master

this was already the case thanks to the protected variables, but now only privileged `deploy` jobs can read  that variable in `master` builds.

Also another useful change is brought by #306: the deploy of `mathcomp/mathcomp:coq-{8.7,8.8,8.9,dev}` will occur only if the `build` stage of math-comp succeeds for these 4 versions of Coq, so this ensure that the version of math-comp available in the 4 Docker images will be consistent anytime. But of course, this deployment condition does not require that the `test` stage succeeds.

@CohenCyril 
> (but I guess it would only prevent mistakes, not deliberate hacking attempts, which could encore the token so that gitlab does not filter it out?)

Indeed, but these hacking attempts should anyway be implemented by adding code in the deploy jobs of the `.gitlab-ci.yml` file… so this would be noticeable for the person that merges the PR.

CohenCyril commented:
> Indeed, but these hacking attempts should anyway be implemented by adding code in the deploy jobs of the `.gitlab-ci.yml` file… so this would be noticeable for the person that merges the PR

@erikmd this assumes that anyone with merge rights is aware of this security breach... and it is not the case yet.

ejgallego commented:
@CohenCyril well people with merge rights should certain refuse to merge if they don't understand what a change in gitlab-ci.yml is doing, right?

CohenCyril commented:
> @CohenCyril well people with merge rights should certain refuse to merge if they don't understand what a change in gitlab-ci.yml is doing, right?

right!
